Change cache getting and setting

rob-b opened the pull request:
The gateway api caches requests made to the other internal api endpoints
but by default cachecontrol caches these requests by their url. We want
to include a user's auth token in the cache key (for private requests)
so that we can cache by user.

When reading from the cache we now always use the authorization header
to build the cache key in conjunction with the request.url. When
actually reading from the backend store we rely on redis' mget method to
request both versions of the key (with and without the hash of the auth
header) and return the first value we find. We are biased towards keys
that do NOT include the auth hash but in practice public/private should
be a binary switch and so there should never be an issue of a less
preferred key being returned.

When writing to the cache we examine the response when creating a cache
key to determine if the response is private or public so we can set an
appropriate cache key

https://thephysicalnetwork.atlassian.net/browse/RO-586

Commit:
Include a hash of the Authorization header in the cache key

This works for privately-cached resources, but it defeates the purpose of the public cache by caching _everything_ per user.

rob-b committed:
Change cache getting and setting

The gateway api caches requests made to the other internal api endpoints
but by default cachecontrol caches these requests by their url. We want
to include a user's auth token in the cache key (for private requests)
so that we can cache by user.

When reading from the cache we now always use the authorization header
to build the cache key in conjunction with the request.url. When
actually reading from the backend store we rely on redis' mget method to
request both versions of the key (with and without the hash of the auth
header) and return the first value we find. We are biased towards keys
that do NOT include the auth hash but in practice public/private should
be a binary switch and so there should never be an issue of a less
preferred key being returned.

When writing to the cache we examine the response when creating a cache
key to determine if the response is private or public so we can set an
appropriate cache key

rob-b commented:
@kirberich I recall you mentioning something about not needing to check the vary header and to use a method but I can't remember what that method was. Something makes me thing you said to use `is_secure` but obv that's not it (since that's a django thing for checking if the request used tls).

kirberich commented on cachecontrol/controller.py:
```diff
@@ -58,8 +59,28 @@ def _urlnorm(cls, uri):
+        return hashlib.sha224(authorization_header.encode('utf-8')).hexdigest()
+
+    @classmethod
+    def cache_key(cls, request, response=None):
```
I don't actually think we need the response (or even an `is_public` flag) anymore because we're using mget.

We can always just create both keys and return which ever one we find in the cache

kirberich commented on cachecontrol/controller.py:
```diff
@@ -58,8 +59,28 @@ def _urlnorm(cls, uri):
+        """
+        normalized_url = cls._urlnorm(request.url)
+        authorization_hash = cls._authorization_digest(request.headers.get('Authorization'))
+        vary_values = response.headers.get('vary', '') if response and response.headers else ''
```
See above, I think this can all go away - also it looks a bit weird that we first create the hash, then check the response headers and then potentially remove the hash again, but this will go away anyway

kirberich commented on cachecontrol/controller.py:
```diff
@@ -90,21 +111,21 @@ def cached_request(self, request):
             return False
 
-        if 'max-age' in cc and cc['max-age'] == 0:
+        if 'max-age' in cc and cc['max-age'] == '0':
```
I think this might be nicer as `int(cc['max-age']) == 0`, just in case we actually do get back an integer for some reason

rob-b replied:
Before fixing this bug I checked how they implement the header parsing and they don't do anything to cast to ints so unless we modify that ourselves only strings will ever come back. That said, max-age is not accessed consistently - it's a combination of this string/int comparison mistake, trying to cast to int and catching the error and an explicit `isdigit` check so I'll make a sweep now and handle it consistently everywhere

kirberich commented on cachecontrol/controller.py:
```diff
@@ -254,9 +275,9 @@ def cache_response(self, request, response, body=None,
-        cache_url = self.cache_url(request.url)
-        logger.debug('Updating cache with response from "%s"', cache_url)
+        cache_key = self.cache_key(request)
+        write_cache_key = self.cache_key(request, response)
```
again as above this can probably go away if we just always create the key with authentication

kirberich commented on cachecontrol/controller.py:
```diff
@@ -317,11 +338,12 @@ def update_cached_response(self, request, response):
-        cache_url = self.cache_url(request.url)
-
+        cache_key = self.cache_key(request)
+        write_cache_key = self.cache_key(request, response)
```
And again here

kirberich reviewed (changes requested):
This looks great! Really just one note on that we can probably get rid of the response in the cache_key method now as we're doing multi-get so it doesn't hurt to look for the auth-key variant even for public requests

rob-b commented:
@kirberich can you take another look?

dotsbb commented on tests/test_storage_redis.py:
```diff
@@ -2,14 +2,61 @@
+    @pytest.mark.parametrize("key, value, expected", [
+        ("key-a", 100, b'100'),
+        ("key-b", 948, b'948'),
+        ("key-b;deadbeef", 3, b'3'),
```
Please add a test case with `value` `0` or empty list/string.

kkorekk commented on cachecontrol/adapter.py:
```diff
@@ -117,8 +117,8 @@ def _update_chunk_length(self):
         if request.method in self.invalidating_methods and resp.ok:
-            cache_url = self.controller.cache_url(request.url)
-            self.cache.delete(cache_url)
+            cache_key = self.controller.cache_key(request)
```
`self.invalidating_methods` shouldn't be extended with `PATCH` ? We're using this method for updating user profile.

kkorekk commented on cachecontrol/caches/redis_cache.py:
```diff
@@ -20,7 +20,13 @@ def __init__(self, conn):
+        # first match
+        key_prefix, _ = keybreaker(key)
+        key = [key_prefix, key]
+        return next((i for i in self.conn.mget(key) if i), None)
```
I think `next` is not required here, `mget` will return None if key doesn't exist so this can be simplified to
```python
values = self.conn.mget([key_prefix, key])
return values[0] or values[1]
```

rob-b commented:
If `mget` returns `None` then

values = self.conn.mget([key_prefix, key])return values[0] or values[1]


Would be an `NoneType is not scriptable` error or a type error (not sure
which). Either way, I think "simpler" is just a matter of taste here

On 27 April 2017 at 12:30, Kornel Maleszka <notifications@github.com> wrote:

> *@kkorekk* commented on this pull request.
> ------------------------------
>
> In cachecontrol/caches/redis_cache.py
> <https://github.com/yourvine/cachecontrol/pull/1#discussion_r113673489>:
>
> > @@ -20,7 +20,13 @@ def __init__(self, conn):
>          self.conn = conn
>
>      def get(self, key):
> -        return self.conn.get(key)
> +        # a key might be of the format foo;bar (where `;` is our default delimter).
> +        # To get this key's value from redis we do `mget foo foo;bar` (this
> +        # returns a list of values for each key searched for) and return the
> +        # first match
> +        key_prefix, _ = keybreaker(key)
> +        key = [key_prefix, key]
> +        return next((i for i in self.conn.mget(key) if i), None)
>
> I think next is not required here, mget will return None if key doesn't
> exist so this can be simplified to
>
> values = self.conn.mget([key_prefix, key])return values[0] or values[1]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yourvine/cachecontrol/pull/1#pullrequestreview-35081242>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFK4l0sD7wLn-vEo63MQv2lAErjoa9eks5r0HxEgaJpZM4NBVU8>
> .
>

kkorekk commented:
@rob-b `mget` will return a list with `None` in place of missing keys, but this is minor comment, so you can ignore it.

rob-b commented:
@kkorekk @dotsbb I just rebased this branch with a few changes; can you take another look.

* added `PATCH` to `invalidating_methods`
* added some tests for `False`-y values

olasitarska commented:
@kirberich are you happy for this to be merged?
